includes/panopto_derivatives.inc

```diff
@@ -5,37 +5,386 @@
  * @param AbstractObject $object An Islandora Object
  * @param String $filePath video file path
  */
+
+
+libraries_load('aws'); 
+
+use Aws\S3\S3Client;
+use Aws\Exception\AwsException;
+use Aws\S3\ObjectUploader;
+use Aws\S3\MultipartUploader;
+use Aws\Exception\MultipartUploadException;
+
+/**
+ * Implement Panopto video upload  
+ * @param $object An Islandora Object
+ * @param $filePath path of object
+ * @return array An array that describes the success or failure of the upload 
+ */
 function islandora_video_streaming_panopto_upload($object, $filePath) {
-  if ($force || !isset[$object['KALTURA_STREAM']) {
+  if ($filePath || !isset($object['PANOPTO_INFO'])) {
+    module_load_include('inc', 'islandora_video', "includes/derivatives" );
     $archival_file = islandora_video_copy_obj_datastream($object);
     if ($archival_file === FALSE) {
       return islandora_video_no_obj_datastream($object->id);
     }
     $filePath = drupal_realpath($archival_file['file']->uri);
     $failureMessage = array(
       'success' => false,
-      'messages' => array(
-        'message' => t('Failed to upload object @pid to Panopto, but no exception caught.'),
+      'messages' => array(),
+    );
+    $streaming = array();
+    
+    $oauthendpoint = variable_get('islandora_video_streaming_panopto_host');
+    $uploadendpoint = $oauthendpoint . '/Panopto/PublicAPI/Rest/sessionUpload';
+    $oauthendpoint .= "/Panopto/oauth2/connect/token";
+    $token = islandora_video_streaming_panopto_get_token($oauthendpoint);
+    $session = null;
+    if ($token) {
+      $session = islandora_video_streaming_panopto_get_session($token, $uploadendpoint); 
+    }
+    $success = false;
+    if ($session) {
+      try {
```

Comment:
This try is really only going to catch exceptions thrown from `islandora_video_streaming_panopto_upload_file`, correct?  Does anything else throw an unhandled exception?

Reply:
I believe so and do you mean outside of the `islandora_video_streaming_panopto_upload_file`?

Reply:
Yes, you have a very broad `try...catch` block, which I think is only effective over the two `islandora_video_streaming_panopto_upload_file` calls in `islandora_video_streaming_panopto_upload`.

Reply:
Since it's only effective over the two functions, should I reduce the try...catch to only consist of the two upload functions?

Reply:
Yep.  You can also get away with starting the `try` early, and ending it just after the last `catch`able statement, but since it is not really doing anything for the earlier statements, it is better to keep it more specifically scoped.